af/fastpath/fastpath/core.py

```diff
@@ -1248,71 +1276,27 @@ def score_http_requests(msm: dict) -> dict:
     tk = g_or(msm, "test_keys", {})
-    requests = tk.get("requests", []) or []
-    for r in requests:
-        is_tor = gn(r, "request", "tor", "is_tor")
-        body = gn(r, "response", "body")
-        if is_tor is None or body is None:
+    requests = g_or(tk, "requests", [])
+    matches: List[Dict] = []
+    for req in requests:
+        is_tor = gn(req, "request", "tor", "is_tor")
+        resp = g_or(req, "response", {})
+        body = gn(req, "response", "body")
+        match_http_body_fingerprints(resp, matches)
+        match_http_headers_fingerprints(resp, matches)
+
+    if matches:
+        if is_tor:
             scores["accuracy"] = 0.0
-            log.debug(f"Incorrect measurement t2 {rid} {inp}")
-            return scores
-
-        if isinstance(body, dict):
-            # TODO: is this needed?
-            if "data" in body and g_or(body, "format", "") == "base64":
-                log.debug("Decoding base64 body")
-                body = b64decode(body["data"])
-
-            else:
-                logbug(3, "incorrect body of type dict", msm)
-                body = None
-
-        for fp in zzfps["body_match"] + g_or(ccfps, "body_match", []):
-            bm = fp["body_match"]
-            if isinstance(body, bytes):
-                idx = body.find(bm.encode())
-            else:
-                idx = body.find(bm)
+            log.debug(f"Failed measurement t1 {rid} {inp}")
 
-            if idx != -1:
-                if is_tor:
-                    scores["accuracy"] = 0.0
-                    log.debug(f"Failed measurement t1 {rid} {inp}")
-                    return scores
-                scores["confirmed"] = True
-                log.debug("matched body fp %s %r at pos %d", msm_cc, bm, idx)
-
-        # Match HTTP headers if found
-        headers = g_or(r, "headers", {})
-        headers = {h.lower(): v for h, v in headers.items()}
-        for fp in zzfps["header_full"] + g_or(ccfps, "header_full", []):
-            name = fp["header_name"]
-            if name in headers and headers[name] == fp["header_full"]:
-                if is_tor:
-                    scores["accuracy"] = 0.0
-                    log.debug(f"Failed measurement t2 {rid} {inp}")
-                    return scores
-                scores["confirmed"] = True
-                log.debug("matched header full fp %s %r", msm_cc, fp["header_full"])
-
-        for fp in zzfps["header_prefix"] + g_or(ccfps, "header_prefix", []):
-            name = fp["header_name"]
-            prefix = fp["header_prefix"]
-            if name in headers and headers[name].startswith(prefix):
-                if is_tor:
-                    scores["accuracy"] = 0.0
-                    log.debug(f"Failed measurement {rid} {inp}")
-                    return scores
-                scores["confirmed"] = True
-                log.debug("matched header prefix %s %r", msm_cc, prefix)
+        else:
+            scores["confirmed"] = True
```

Comment:
This should be set only when `probe_cc` of the measurement is in the `expected_countries` of any matched fingerprint OR scope in any of the matching fingerprints is one of `[isp, nat]`

Reply:
This code seems to be duplicate of what is here: https://github.com/ooni/pipeline/pull/407/files#r1118514733. It's probably best to refactor this into a dedicated function so if we make changes to the logic (as we improve our understanding of how accurate the fingerprints are), we can update it only in one place and not forget the other.

Reply:
Part of the code is duplicated on purpose in order to merge it later on based on how it evolves in the meantime. Part of it has been deduplicated into match_http_body_fingerprints and match_http_headers_fingerprints